diesel/src/sqlite/mod.rs

```diff
@@ -13,3 +13,21 @@ pub mod query_builder;
 pub use self::backend::{Sqlite, SqliteType};
 pub use self::connection::SqliteConnection;
 pub use self::query_builder::SqliteQueryBuilder;
+
+/// Trait for the implementation of a SQLite aggregate function
+///
+/// This trait is to be used in conjunction with the `sql_function!`
+/// macro for defining a custom SQLite aggregate function. See
+/// the documentation [there](../prelude/macro.sql_function.html) for details.
+pub trait SqliteAggregateFunction<Args>: Default {
+    /// The result type of the SQLite aggregate function
+    type Output;
+
+    /// The `step()` method is called once for every record of the query
+    fn step(&mut self, args: Args);
+    /// After the last row has been processed, the `finalize()` method is
+    /// called to compute the result of the aggregate function. If no rows
+    /// were processed, `finalize()` is not called and `None` is returned
+    /// instead
+    fn finalize(self) -> Self::Output;
```

Comment:
I had recently another thought about this, I think explicitly returning `Option<Self::Output>` here as actually the better solution, as there could be reasons not to return `NULL` for an empty set. A trivial example is the `count(something)` sql function, that returns `0` for empty sets. Defining such a function would be not possible with the current implementation. Based on that I propose the following change:

```suggestion
    fn finalize(aggregator: Option<Self>) -> Option<Self::Output>;
```

this allows that users set a custom default value or just return `None` for the `None` case by using the `?` operator on the `aggregator`

Reply:
Makes sense to me! I'll try to get to that this weekend.

Reply:
@weiznich I ran into some trouble with types. Please take a close look at commit [44ffaa4](https://github.com/diesel-rs/diesel/pull/2325/commits/44ffaa46b780ff90c9b91f4e0e6b442d0e27c762). The change I'm most curious about is in `src/type_impls/option.ts`. This change was needed so I could pass `Option<Ret>` to `process_sql_function_result`, since at the time, only `Nullable` implemented `ToSql` for `Option`, but there wasn't a way for me to tell the compiler that `RetSqlType` is `Nullable`. If I don't make this change, I would need to do this in `src/sqlite/connection/raw.ts` in the `run_aggregator_final_function`:

```rust
let result = A::finalize(aggregator);

match result {
  Some(r) => match process_sql_function_result::<RetSqlType, Ret>(r) {
    ...
  },
  None => {
    // The same code that exists in process_sql_function_result
  }
}
```

This felt like something I shouldn't have to do, but I was running into compile errors. I think the change I made is safe because only `Nullable<ST>` satisfies `T: IntoNullable<Nullable = T>`. Please let me know if there is something or some approach I am missing!

Reply:
I think it should be possible to define a `count` function as following:

```rust
sql_function!(#[aggregate] fn count(x: Integer) -> Integer);
```

This implies that my proposed change does not completely make sense. We need to return `Self::Output` instead of `Option<Self::Output>` as we may want to return a plain value. 

I've pushed this as https://github.com/diesel-rs/diesel/pull/2325/commits/5ffc26bb5e4fa84ecd3899238ff82e158add3a6c to this branch. Would be great if you could have a final look at that change, otherwise I think we are finished here. I will also try to do a final review in the next few days and after that merge this PR.

Reply:
This looks good to me! Thanks for all your help 🎉